### PR TITLE
bug fixed leaky_relu and negative testing.

### PR DESCRIPTION
**Pull request by kartik4949**

### CODE REVIEW

**Comment by kartik4949** [comment]
adding negative tests for all tests

**Comment by kartik4949** [comment]
@geohot added neg tests

**Comment by kartik4949** [comment]
more clean now
@geohot @marcelbischoff what you say.

**Comment by kartik4949** [comment]
@geohot @marcelbischoff 
tests failing for negative values
conv
test_div (because of np.log(x))

**Comment by kartik4949** [comment]
conv and tensordot needs lot of tolerance  around 1-e3

**Comment by kartik4949** [comment]
@geohot PR completed
Added leaky_relu fix
Added Negative Testing 
Adde __neg__ support.

File: tinygrad/tensor.py [hunk 1]
```diff
@@ -245,6 +249,9 @@ def dispatch(*x, **kwargs):
     return f.apply(f, *x, **kwargs)
   setattr(Tensor, name, dispatch)
   # TODO: div is a second class op, so it doesn't work here
+  if not hasattr(Tensor, "__neg__"):
```
Comment (geohot) [inline tinygrad/tensor.py:252]: Indentation. And this runs for every name.

File: tinygrad/tensor.py [hunk 2]
```diff
@@ -200,11 +200,15 @@ def swish(self):
...
     return 2.0 * ((2.0 * self).sigmoid()) - 1.0
 
-  def leakyrelu(self, neg_slope=0.01):
-    return self.relu() + (-neg_slope*self).relu()
+  def leaky_relu(self, neg_slope=0.01):
+    return self.relu() - (-neg_slope*self).relu()
 
   def abs(self):
-    return self.relu() + (-1.0*self).relu()
+    return self.relu() + (-*self).relu()
```
Comment (geohot) [inline tinygrad/tensor.py:207]: Is that a real op?

File: tinygrad/ops_cpu.py [hunk 3]
```diff
@@ -157,9 +159,9 @@ class LogSoftmax(Function):
   @staticmethod
   def forward(ctx, input):
     def logsumexp(x):
-      #return np.log(np.exp(x).sum(axis=1))
-      c = x.max(axis=1)
-      return c + np.log(np.exp(x-c.reshape((-1, 1))).sum(axis=1))
+      with np.errstate(invalid='ignore'):
```
Comment (geohot) [inline tinygrad/ops_cpu.py:162]: Why do we need these?

File: tinygrad/tensor.py (hunk 2 above)
Comment (kartik4949) [reply tinygrad/tensor.py:207]: mistake haha * needs to be removed

File: tinygrad/tensor.py (hunk 1 above)
Comment (kartik4949) [reply tinygrad/tensor.py:252]: uhm.. but will add __neg__ once..

File: tinygrad/ops_cpu.py [hunk 4]
```diff
@@ -157,9 +159,9 @@ class LogSoftmax(Function):
...
+        c = x.max(axis=1)
+        return c + np.log(np.exp(x-c.reshape((-1, 1))).sum(axis=1))
```
[the remark this answers is not in the document]
Comment (geohot) [reply tinygrad/ops_cpu.py:164]: That shouldn't happen, that's undefined.

**Comment by kartik4949** [comment]
closing this will add '__neg__' support in future pr.

**Pull request closed without merging**